src/pytest_mock_resources/fixture/database/relational/postgresql.py

```diff
@@ -29,27 +32,40 @@ def create_postgres_fixture(*ordered_actions, **kwargs):
 
     session = kwargs.pop("session", None)
 
+    async_ = kwargs.pop("async_", False)
+
+    if async_ and session and not isinstance(session, AsyncSession):
+        raise ValueError("When async_ is `True`, session must be an instance of `AsyncSession`")
+
     if len(kwargs):
         raise KeyError("Unsupported Arguments: {}".format(kwargs))
 
     @pytest.fixture(scope=scope)
     def _(_postgres_container, pmr_postgres_config):
         database_name = _create_clean_database(pmr_postgres_config)
```

Comment:
i would suggest maybe something like
```
def create_engine_manager(pmr_postgres_config):
    database_name = _create_clean_database(pmr_postgres_config)
    engine = get_sqlalchemy_engine(pmr_postgres_config, database_name)
    assign_fixture_credentials(...)
    return EngineManager(engine, ordered_actions, tables=tables, default_schema="public")

def create_postgres_fixture(*ordered_actions, **kwargs):
    ...

    @pytest.fixture(scope=scope)
    def _sync(_postgres_container, pmr_postgres_config):
        engine_manager = create_engine_manager(pmr_postgres_config)
        yield from engine_maneger.manage_sync(session=session)

    @pytest.fixture(scope=scope)
    def _async(_postgres_container, pmr_postgres_config):
        engine_manager = create_engine_manager(pmr_postgres_config)
        yield from engine_maneger.manage_async(session=session)

    if async_:
        return _async
    else:
        return _sync
```

This way, you're in the context of an async method when performing the async session context management, which should mean you wouldn't actually need direct access to a `loop` object to run `loop.run_until_complete`.

and you might not even have to have `pytest-asyncio` as a direct dependency (like i suggested above) anymore either. Technically we don't need it, and practically, it might be preferable to not **require** it because, for example, `pytest-asyncio` and `pytest-aiohttp`('s impl of similar async fixture support things) are incompatible with one another.

(although practically, in order to get a handle on a loop in order to run async fixtures, you **will** ultimately need pytest-asyncio or something like it, so that'd be something to mention in the docs anyways.

Reply:
Hey @DanCardin, I went with a separate implementation because of the internal workings of `manage(session)` specifically `self._run_actions()`. That logic is incompatible with an async engine/session. If we move the call to `self._run_actions()` outside of the `manage...()` method then the implementation you have proposed will be feasible.

Reply:
Also, we would have to consider the session being passed into the `manage()` method i.e whether it's an `AsyncSession` or a normal session.

Reply:
My larger point is that the easiest way forward would be performing all the internal bits of sqlalchemy interactions with the sync engine, and just construct/yield the async engine to the test at the end.

Endgame is probably making this whole process more agnostic to the engine in the first place so there can be a single intercompatible implementation. but until that point, it doesn't seem obvious to me why the test should care whether the internals of the plugin are sync or not. I assume all **you** care about is that your test is being handed an async engine

Reply:
Oh, I understand. Let me do some refactoring.

Reply:
@DanCardin I've made changes. Do you mind taking a look? One thing I noticed is that ordered actions that have `create temp` DDL statement won't work because the sync and async engines have different sessions.

Reply:
That makes sense. Given how databases get provisioned, i think there's no real practical reason why one would `CREATE TEMP` though. each test is granted a unique new database, so tests cannot conflict anyways